lib/rules.js

```diff
@@ -0,0 +1,280 @@
+          }
+        }
+      }
+    }
+
+    this.conditions = conditions;
+  }
+
+  /**
+   * Throw errors from these conditions so they can be traced back
+   * to the rule they came from
+   */
+  _throw(code, msg) {
+    let err = new Error(msg);
+    err.code = code;
+    throw err;
+  }
+
+  /**
+   * Compare the specified value to the required value.  This is its own
+   * function to make it easier to replace the comparison system, e.g.
+   * switching from globbing to regular expressions
+   *
+   * Return true if the value satisfies the specification
+   */
+  _compare(condition, satisfier) {
+    return condition === satisfier;
+  }
+
+  // Evaluate a specific condition of the conditions against the value
+  // provided
+  _evaluateCondition(condition, satisfier) {
+    if (typeof condition === 'string') {
+      return this._compare(condition, satisfier);
+    }
+
+    // condition can be a single item or a list which makes naming icky. Think
+    // as if this were "let condition of conditions"
+    for (let aCondition of condition) {
+      if (this._compare(aCondition, satisfier)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Evaluate all provided conditions.  Return true if the conditions are
+   * satisfied or false if any condition is not satisified
+   */
+  evaluate(satisfiers) {
+    if (this.conditions === null) {
+      return true;
+    }
+
+    if (typeof satisfiers !== 'object') {
+      this._throw(ErrorCodes.InvalidSatisfiers, 'satisfiers must be an object');
+    }
+
+    for (let condition of Object.keys(this.conditions)) {
+      if (!satisfiers[condition]) {
+        this._throw(ErrorCodes.InvalidSatisfiers, `required satisfier for condition ${condition} not specified`);
+      }
+      if (!this._evaluateCondition(this.conditions[condition], satisfiers[condition])) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Return a list of satisfier keys which are required to evaluate this
+   * ruleset
+   */
+  requiredSatisfiers() {
+    return Object.keys(this.conditions);
+  }
+}
+
+/**
+ * Rule encapsulates how a rule is to be evaluated.
+ */
+class Rule {
+  constructor({ruleId, conditions, values, description}) {
+    if (typeof ruleId !== 'string') {
+      this._throw(ErrorCodes.InvalidRule, 'ruleId must be string');
+    }
+    this.ruleId = ruleId
+
+    if (typeof conditions !== 'object') {
+      this._throw(ErrorCodes.InvalidConditions, 'conditions must be an object');
+    }
+    this.conditions = new Conditions(conditions, this);
+
+    if (typeof values !== 'object') {
+      this._throw(ErrorCodes.InvalidValues, 'values must be an object');
+    }
+    this.values = values;
+
+    if (typeof description !== 'string') {
+      this._throw(ErrorCodes.InvalidRule, 'description must be a string');
+    }
+  }
+
+  _throw(code, msg) {
+    let err = new Error(msg);
+    err.code = code;
+    err.ruleId = this.ruleId || '<unknown>';
```

Comment:
What if we just enforce being given a roleId here and in `Conditions`?

Reply:
Because `_throw` is also being used to throw when a ruleId is not passed to the contructor.  In condtions, the check is to allow creation of Conditions without a rule, basically unit tests.